File: js/maximumUnits.js
/**
 * @param {number[][]} boxTypes
 * @param {number} truckSize
 * @return {number}
 */
var maximumUnits = function (boxTypes, truckSize) {
    let boxes = 0

    boxTypes.sort(
        (function (index) {
            return function (a, b) {
                return a[index] === b[index] ? 0 : a[index] > b[index] ? -1 : 1
            }
        })(1)
    )

    if (truckSize == 1) return boxTypes[0][1]

    for (let box of boxTypes) {
        if (truckSize >= box[0]) {
            boxes += box[0] * box[1]
            truckSize -= box[0]
        } else {
            boxes += truckSize * box[1]
            return boxes
        }
    }
}

console.log(
    maximumUnits(
        [
            [1, 3],
            [5, 5],
            [2, 5],
            [4, 2],
            [4, 1],
            [3, 1],
            [2, 2],
            [1, 3],
            [2, 5],
            [3, 2]
        ],
        35
    )
)

File: Python/maximumUnits.py
class Solution:
    def maximumUnits(self, boxTypes: List[List[int]], truckSize: int) -> int:
        boxes, cur_units, cnt = 0, 1000, Counter()
        for box, units in boxTypes:
            cnt[units] += box
        while cur_units > 0:
            if cnt[cur_units] > 0:
                fit_in = min(truckSize, cnt[cur_units])
                boxes += fit_in * cur_units
                truckSize -= fit_in
                if truckSize == 0:
                    return boxes
            cur_units -= 1
        return boxes
